File: apps/digital-health-startup/src/features/chat/services/agent-selector-service.ts
/**
 * AGENT SELECTOR SERVICE
 * 
 * Shared utilities for intelligent agent selection in Mode 2.
 * Provides query analysis, agent search, and ranking capabilities.
 * 
 * Features:
 * - Query analysis using OpenAI to extract intent and domains
 * - GraphRAG hybrid search for candidate agents (Pinecone + Supabase)
 * - Multi-criteria agent ranking (similarity, domain, tier, capabilities)
 * - Confidence scoring and reasoning generation
 * - Enterprise-grade observability and error handling
 */

import { createClient } from '@supabase/supabase-js';
import { Pinecone } from '@pinecone-database/pinecone';
import { createLogger } from '@/lib/services/observability/structured-logger';
import {
  AgentSelectionError,
  PineconeConnectionError,
  GraphRAGSearchError,
  AgentNotFoundError,
} from '@/lib/errors/agent-errors';
import {
  validateAgentSearchQuery,
  safeValidate,
  getValidationErrors,
  AgentSearchQuerySchema,
} from '@/lib/validators/user-agents-schema';
import { agentGraphRAGService } from '@/lib/services/agents/agent-graphrag-service';
import type { AgentSearchResult } from '@/lib/services/agents/agent-graphrag-service';

// ============================================================================
// TYPES
// ============================================================================

export interface Agent {
  id: string;
  name: string;
  display_name: string;
  description: string;
  system_prompt: string;
  tier: number;
  capabilities?: string[];
  knowledge_domains?: string[];
  specialties?: string[];
  model?: string;
  metadata?: any;
}

export interface QueryAnalysis {
  intent: string;
  domains: string[];
  complexity: 'low' | 'medium' | 'high';
  keywords: string[];
  medicalTerms: string[];
  confidence: number;
}

export interface AgentRanking {
  agent: Agent;
  score: number;
  reason: string;
  breakdown: {
    semanticSimilarity: number;
    domainRelevance: number;
    tierPreference: number;
    capabilityMatch: number;
  };
}

export interface AgentSelectionResult {
  selectedAgent: Agent;
  confidence: number;
  reasoning: string;
  alternativeAgents: AgentRanking[];
  analysis: QueryAnalysis;
}

// ============================================================================
// AGENT SELECTOR SERVICE CLASS
// ============================================================================

export class AgentSelectorService {
  private supabase;
  private pinecone: Pinecone;
  private openaiApiKey: string;
  private logger;

  constructor(options?: { requestId?: string; userId?: string }) {
    // Initialize Supabase client
    const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!;
    const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY!;
    this.supabase = createClient(supabaseUrl, supabaseKey);

    // Initialize Pinecone client
    this.pinecone = new Pinecone({
      apiKey: process.env.PINECONE_API_KEY!,
    });

    this.openaiApiKey = process.env.OPENAI_API_KEY!;

    // Initialize structured logger
    this.logger = createLogger({
      requestId: options?.requestId,
      userId: options?.userId,
    });
  }

  /**
   * Analyze query to extract intent, domains, and complexity
   * Uses OpenAI to classify medical query characteristics
   */
  async analyzeQuery(query: string): Promise<QueryAnalysis> {
    const operationId = `analysis_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const startTime = Date.now();

    this.logger.info('query_analysis_started', {
      operation: 'analyzeQuery',
      operationId,
      queryLength: query.length,
      queryPreview: query.substring(0, 100),
    });

    try {
      const response = await fetch('https://api.openai.com/v1/chat/completions', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.openaiApiKey}`
        },
        body: JSON.stringify({
          model: 'gpt-4-turbo-preview',
          messages: [{
            role: 'system',
            content: `Analyze this medical/healthcare query and extract structured information. 
            
            Return a JSON object with:
            - intent: Primary intent (diagnosis, treatment, research, consultation, education, etc.)
            - domains: Array of medical domains (cardiology, oncology, neurology, etc.)
            - complexity: Complexity level (low, medium, high)
            - keywords: Key medical terms and concepts
            - medicalTerms: Specific medical terminology
            - confidence: Confidence score (0-1) in your analysis
            
            Focus on medical/healthcare context. If not medical, still analyze for general intent.`
          }, {
            role: 'user',
            content: query
          }],
          response_format: { type: 'json_object' },
          temperature: 0.3
        })
      });

      const data = await response.json();
      const analysis = JSON.parse(data.choices[0].message.content);

      console.log(`✅ [AgentSelector] Query analysis complete:`, {
        intent: analysis.intent,
        domains: analysis.domains,
        complexity: analysis.complexity,
        confidence: analysis.confidence
      });

      return {
        intent: analysis.intent || 'general',
        domains: analysis.domains || [],
        complexity: analysis.complexity || 'medium',
        keywords: analysis.keywords || [],
        medicalTerms: analysis.medicalTerms || [],
        confidence: analysis.confidence || 0.7
      };

    } catch (error) {
      console.error('❌ [AgentSelector] Query analysis failed:', error);
      
      // Fallback analysis
      return {
        intent: 'general',
        domains: [],
        complexity: 'medium',
        keywords: query.toLowerCase().split(' ').slice(0, 5),
        medicalTerms: [],
        confidence: 0.5
      };
    }
  }

  /**
   * Search for candidate agents using GraphRAG hybrid search
   * Falls back to database search if GraphRAG fails
   * 
   * Uses enterprise-grade:
   * - Zod validation
   * - Structured logging
   * - Typed error handling
   * - Performance metrics
   */
  async findCandidateAgents(
    query: string,
    domains: string[],
    limit: number = 10
  ): Promise<Agent[]> {
    const operationId = `search_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const startTime = Date.now();

    this.logger.info('agent_search_started', {
      operation: 'findCandidateAgents',
      operationId,
      queryLength: query.length,
      queryPreview: query.substring(0, 100),
      domains,
      limit,
    });

    try {
      // Validate inputs with Zod
      const queryValidation = safeValidate(
        AgentSearchQuerySchema,
        {
          query,
          topK: limit,
          minSimilarity: 0.6,
          filters: {
            knowledge_domain: domains.length > 0 ? domains[0] : undefined,
            status: 'active',
          },
        }
      );

      if (!queryValidation.success) {
        const errors = getValidationErrors(queryValidation.errors);
        this.logger.warn('agent_search_validation_failed', {
          operationId,
          errors,
        });
        // Continue with defaults, but log the issue
      }

      // Step 1: Try GraphRAG hybrid search first (preferred method)
      try {
        const graphRAGResults = await agentGraphRAGService.searchAgents({
          query,
          topK: limit,
          minSimilarity: 0.6,
          filters: {
            knowledge_domain: domains.length > 0 ? domains[0] : undefined,
            status: 'active',
          },
        });

        const agents = graphRAGResults.map((result: AgentSearchResult) => result.agent);

        const duration = Date.now() - startTime;
        this.logger.infoWithMetrics('agent_search_completed', duration, {
          operation: 'findCandidateAgents',
          operationId,
          method: 'graphrag_hybrid',
          resultCount: agents.length,
          topSimilarity: graphRAGResults[0]?.similarity,
        });

        return agents;
      } catch (graphRAGError) {
        // GraphRAG failed, fall back to database search
        this.logger.warn('graphrag_search_failed', {
          operationId,
          error: graphRAGError instanceof Error ? graphRAGError.message : String(graphRAGError),
          fallback: 'database',
        });

        throw new GraphRAGSearchError(
          'GraphRAG search failed, falling back to database',
          {
            cause: graphRAGError instanceof Error ? graphRAGError : new Error(String(graphRAGError)),
            context: { operationId, query: query.substring(0, 100), domains },
          }
        );
      }
    } catch (error) {
      // If it's a GraphRAG error, fall back to database
      if (error instanceof GraphRAGSearchError) {
        return await this.fallbackAgentSearch(query, domains, limit, operationId);
      }

      // For other errors, log and fall back
      this.logger.error(
        'agent_search_error',
        error instanceof Error ? error : new Error(String(error)),
        {
          operation: 'findCandidateAgents',
          operationId,
          query: query.substring(0, 100),
          domains,
        }
      );

      return await this.fallbackAgentSearch(query, domains, limit, operationId);
    }
  }

  /**
   * Fallback agent search using database text search
   */
  private async fallbackAgentSearch(
    query: string,
    domains: string[],
    limit: number
  ): Promise<Agent[]> {
    console.log('🔄 [AgentSelector] Using fallback database search...');

    try {
      // Build optimized query with proper filtering
      let queryBuilder = this.supabase
        .from('agents')
        .select(`
          id,
          name,
          description,
          system_prompt,
          tier,
          capabilities,
          knowledge_domains,
          specialties,
          model,
          metadata
        `);

      // Add domain filtering if available
      if (domains.length > 0) {
        queryBuilder = queryBuilder.overlaps('knowledge_domains', domains);
      }

      // Order by tier and priority (higher tier = better)
      queryBuilder = queryBuilder
        .order('tier', { ascending: true })
        .limit(limit * 2); // Get more candidates for filtering

      const { data: agents, error } = await queryBuilder;

      if (error) {
        console.error('❌ [AgentSelector] Fallback search failed:', error);
        // Return any available agent as last resort
        return await this.getAnyAvailableAgent(limit);
      }

      // If no domain-specific results, relax filters
      if (!agents || agents.length === 0) {
        console.log('⚠️ [AgentSelector] No domain-specific agents found, using general fallback');
        
        const { data: generalAgents } = await this.supabase
          .from('agents')
          .select('*')
          .order('tier', { ascending: true })
          .limit(limit);
        
        console.log(`✅ [AgentSelector] General fallback found ${generalAgents?.length || 0} agents`);
        return generalAgents || [];
      }

      console.log(`✅ [AgentSelector] Fallback search found ${agents.length} agents`);
      return agents;

    } catch (error) {
      console.error('❌ [AgentSelector] Fallback search error:', error);
      // Return any available agent as last resort
      return await this.getAnyAvailableAgent(limit);
    }
  }

  /**
   * Get any available agent as a last resort fallback
   */
  private async getAnyAvailableAgent(limit: number = 5): Promise<Agent[]> {
    console.log('🔄 [AgentSelector] Getting any available agent as emergency fallback...');
    
    try {
      const { data: agents, error } = await this.supabase
        .from('agents')
        .select('*')
        .order('tier', { ascending: true })
        .order('priority', { ascending: false })
        .limit(limit);

      if (error) {
        console.error('❌ [AgentSelector] Failed to get any agent:', error);
        return [];
      }

      console.log(`✅ [AgentSelector] Found ${agents?.length || 0} agents as fallback`);
      return agents || [];
    } catch (error) {
      console.error('❌ [AgentSelector] Error getting fallback agents:', error);
      return [];
    }
  }

  /**
   * Rank agents based on multiple criteria
   * - Semantic similarity to query
   * - Domain relevance
   * - Tier level (higher tier preferred)
   * - Capability match
   */
  rankAgents(
    agents: Agent[],
    query: string,
    analysis: QueryAnalysis
  ): AgentRanking[] {
    console.log(`📊 [AgentSelector] Ranking ${agents.length} agents...`);

    const rankings: AgentRanking[] = agents.map(agent => {
      const breakdown = this.calculateScoreBreakdown(agent, query, analysis);
      const totalScore = this.calculateTotalScore(breakdown);
      const reason = this.generateRankingReason(agent, breakdown, analysis);

      return {
        agent,
        score: totalScore,
        reason,
        breakdown
      };
    });

    // Sort by score (highest first)
    rankings.sort((a, b) => b.score - a.score);

    console.log(`✅ [AgentSelector] Agent ranking complete. Top agent: ${rankings[0]?.agent.name} (score: ${rankings[0]?.score.toFixed(3)})`);

    return rankings;
  }

  /**
   * Calculate detailed score breakdown for an agent
   */
  private calculateScoreBreakdown(
    agent: Agent,
    query: string,
    analysis: QueryAnalysis
  ): AgentRanking['breakdown'] {
    // 1. Semantic similarity (based on description and capabilities)
    const semanticSimilarity = this.calculateSemanticSimilarity(agent, query);

    // 2. Domain relevance
    const domainRelevance = this.calculateDomainRelevance(agent, analysis.domains);

    // 3. Tier preference (higher tier = higher score)
    const tierPreference = this.calculateTierPreference(agent.tier);

    // 4. Capability match
    const capabilityMatch = this.calculateCapabilityMatch(agent, analysis);

    return {
      semanticSimilarity,
      domainRelevance,
      tierPreference,
      capabilityMatch
    };
  }

  /**
   * Calculate semantic similarity based on text overlap
   */
  private calculateSemanticSimilarity(agent: Agent, query: string): number {
    const queryTerms = query.toLowerCase().split(' ').filter(term => term.length > 2);
    const agentText = `${agent.name} ${agent.description} ${agent.system_prompt}`.toLowerCase();
    
    const matchingTerms = queryTerms.filter(term => agentText.includes(term));
    return matchingTerms.length / Math.max(queryTerms.length, 1);
  }

  /**
   * Calculate domain relevance score
   */
  private calculateDomainRelevance(agent: Agent, queryDomains: string[]): number {
    if (queryDomains.length === 0) return 0.5; // Neutral if no domains specified
    
    const agentDomains = agent.knowledge_domains || [];
    const matchingDomains = queryDomains.filter(domain => 
      agentDomains.some(agentDomain => 
        agentDomain.toLowerCase().includes(domain.toLowerCase()) ||
        domain.toLowerCase().includes(agentDomain.toLowerCase())
      )
    );
    
    return matchingDomains.length / queryDomains.length;
  }

  /**
   * Calculate tier preference score
   */
  private calculateTierPreference(tier: number): number {
    // Normalize tier to 0-1 scale (assuming tiers 1-5)
    return Math.min(tier / 5, 1);
  }

  /**
   * Calculate capability match score
   */
  private calculateCapabilityMatch(agent: Agent, analysis: QueryAnalysis): number {
    const agentCapabilities = agent.capabilities || [];
    const queryKeywords = [...analysis.keywords, ...analysis.medicalTerms];
    
    if (queryKeywords.length === 0) return 0.5; // Neutral if no keywords
    
    const matchingCapabilities = queryKeywords.filter(keyword =>
      agentCapabilities.some(capability =>
        capability.toLowerCase().includes(keyword.toLowerCase()) ||
        keyword.toLowerCase().includes(capability.toLowerCase())
      )
    );
    
    return matchingCapabilities.length / queryKeywords.length;
  }

  /**
   * Calculate total weighted score
   */
  private calculateTotalScore(breakdown: AgentRanking['breakdown']): number {
    // Weighted combination of all factors
    const weights = {
      semanticSimilarity: 0.4,
      domainRelevance: 0.3,
      tierPreference: 0.2,
      capabilityMatch: 0.1
    };

    return (
      breakdown.semanticSimilarity * weights.semanticSimilarity +
      breakdown.domainRelevance * weights.domainRelevance +
      breakdown.tierPreference * weights.tierPreference +
      breakdown.capabilityMatch * weights.capabilityMatch
    );
  }

  /**
   * Generate human-readable ranking reason
   */
  private generateRankingReason(
    agent: Agent,
    breakdown: AgentRanking['breakdown'],
    analysis: QueryAnalysis
  ): string {
    const reasons: string[] = [];

    if (breakdown.domainRelevance > 0.7) {
      reasons.push(`expertise in ${analysis.domains.join(', ')}`);
    }

    if (breakdown.semanticSimilarity > 0.6) {
      reasons.push('highly relevant to your query');
    }

    if (breakdown.tierPreference > 0.8) {
      reasons.push('high-tier specialist');
    }

    if (breakdown.capabilityMatch > 0.5) {
      reasons.push('matching capabilities');
    }

    if (reasons.length === 0) {
      return 'general purpose agent';
    }

    return reasons.join(', ');
  }

  /**
   * Generate embedding for text using OpenAI
   */
  private async generateEmbedding(text: string): Promise<number[]> {
    try {
      const response = await fetch('https://api.openai.com/v1/embeddings', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.openaiApiKey}`
        },
        body: JSON.stringify({
          input: text,
          model: 'text-embedding-3-small'
        })
      });

      const data = await response.json();
      return data.data[0].embedding;

    } catch (error) {
      console.error('❌ [AgentSelector] Embedding generation failed:', error);
      throw error;
    }
  }

  /**
   * Complete agent selection workflow
   * Combines query analysis, agent search, and ranking
   */
  async selectBestAgent(query: string): Promise<AgentSelectionResult> {
    console.log('🎯 [AgentSelector] Starting complete agent selection workflow...');

    try {
      // Step 1: Analyze query
      const analysis = await this.analyzeQuery(query);

      // Step 2: Find candidate agents
      const candidates = await this.findCandidateAgents(query, analysis.domains, 10);

      if (candidates.length === 0) {
        throw new Error('No agents found for the given query');
      }

      // Step 3: Rank agents
      const rankings = this.rankAgents(candidates, query, analysis);

      // Step 4: Select best agent
      const selectedAgent = rankings[0];
      const confidence = Math.min(selectedAgent.score + analysis.confidence, 1);

      console.log(`✅ [AgentSelector] Agent selection complete:`, {
        selectedAgent: selectedAgent.agent.name,
        confidence: confidence.toFixed(3),
        totalCandidates: candidates.length
      });

      return {
        selectedAgent: selectedAgent.agent,
        confidence,
        reasoning: selectedAgent.reason,
        alternativeAgents: rankings.slice(1, 4), // Top 3 alternatives
        analysis
      };

    } catch (error) {
      console.error('❌ [AgentSelector] Agent selection failed:', error);
      throw error;
    }
  }
}

// ============================================================================
// EXPORT
// ============================================================================

export const agentSelectorService = new AgentSelectorService();

File: services/ai-engine/src/services/agent_selector_service.py
"""
Agent Selector Service for VITAL Path
Enterprise-grade agent selection with query analysis and intelligent matching

This service provides:
- Query analysis using LLM (intent, domains, complexity, keywords)
- Agent matching using embeddings and domain filtering
- Multi-criteria ranking (similarity, domain, tier, capabilities)
- Comprehensive error handling and observability

Following industry best practices:
- SOLID principles (Single Responsibility, Dependency Inversion)
- Clean Architecture with proper abstractions
- Structured logging and metrics
- Type safety with Pydantic models
- Error classification and handling
"""

import json
from typing import Dict, Any, Optional, List
from datetime import datetime
import structlog
from openai import OpenAI
from pydantic import BaseModel, Field

from core.config import get_settings
from services.supabase_client import SupabaseClient
from services.graphrag_selector import GraphRAGSelector, get_graphrag_selector

logger = structlog.get_logger()
settings = get_settings()


# ============================================================================
# REQUEST/RESPONSE MODELS
# ============================================================================

class QueryAnalysisRequest(BaseModel):
    """Request model for query analysis"""
    query: str = Field(..., min_length=1, max_length=2000, description="User query to analyze")
    user_id: Optional[str] = Field(None, description="User identifier for context")
    tenant_id: Optional[str] = Field(None, description="Tenant/organization identifier")
    correlation_id: Optional[str] = Field(None, description="Correlation ID for tracing")


class QueryAnalysisResponse(BaseModel):
    """Response model for query analysis"""
    intent: str = Field(..., description="Primary intent (diagnosis, treatment, research, consultation, education, etc.)")
    domains: List[str] = Field(..., description="Array of medical domains (cardiology, oncology, neurology, etc.)")
    complexity: str = Field(..., description="Complexity level (low, medium, high)")
    keywords: List[str] = Field(..., description="Key medical terms and concepts")
    medical_terms: List[str] = Field(..., description="Specific medical terminology")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Confidence score (0-1) in the analysis")
    correlation_id: Optional[str] = Field(None, description="Correlation ID for tracing")


class AgentSelectionRequest(BaseModel):
    """Request model for complete agent selection"""
    query: str = Field(..., min_length=1, max_length=2000, description="User query")
    user_id: Optional[str] = Field(None, description="User identifier")
    tenant_id: Optional[str] = Field(None, description="Tenant/organization identifier")
    correlation_id: Optional[str] = Field(None, description="Correlation ID for tracing")
    max_candidates: int = Field(10, ge=1, le=50, description="Maximum candidate agents to consider")


# ============================================================================
# AGENT SELECTOR SERVICE
# ============================================================================

class AgentSelectorService:
    """
    Enterprise-grade agent selector service
    
    Features:
    - LLM-powered query analysis
    - Embedding-based agent matching
    - Multi-criteria ranking
    - Comprehensive error handling
    - Structured logging and metrics
    """
    
    def __init__(
        self,
        openai_client: Optional[OpenAI] = None,
        supabase_client: Optional[SupabaseClient] = None,
        graphrag_selector: Optional[GraphRAGSelector] = None,
        use_graphrag: bool = True
    ):
        """
        Initialize agent selector service

        Args:
            openai_client: OpenAI client instance (for dependency injection)
            supabase_client: Supabase client instance (for dependency injection)
            graphrag_selector: GraphRAG selector for 3-method hybrid selection
            use_graphrag: Enable GraphRAG hybrid selection (default True)
        """
        self.settings = get_settings()
        self.openai_client = openai_client or OpenAI(api_key=self.settings.openai_api_key)
        self.supabase_client = supabase_client
        self.use_graphrag = use_graphrag
        # Pass supabase_client to GraphRAGSelector to ensure it has a properly initialized DB client
        self.graphrag_selector = graphrag_selector or (get_graphrag_selector(supabase_client=supabase_client) if use_graphrag else None)
        
    async def analyze_query(
        self,
        query: str,
        correlation_id: Optional[str] = None
    ) -> QueryAnalysisResponse:
        """
        Analyze query to extract intent, domains, complexity, and keywords
        
        Uses OpenAI GPT-4 for structured query analysis following medical/healthcare context.
        
        Args:
            query: User query text
            correlation_id: Optional correlation ID for tracing
            
        Returns:
            QueryAnalysisResponse with extracted metadata
            
        Raises:
            ValueError: If query is invalid
            Exception: If LLM analysis fails (with fallback)
        """
        operation_id = f"analysis_{datetime.now().timestamp()}_{hash(query) % 10000}"
        start_time = datetime.now()
        
        logger.info(
            "query_analysis_started",
            operation="analyzeQuery",
            operation_id=operation_id,
            correlation_id=correlation_id,
            query_length=len(query),
            query_preview=query[:100]
        )
        
        try:
            # Validate input
            if not query or not query.strip():
                raise ValueError("Query cannot be empty")
            
            # Prepare system prompt for structured analysis
            system_prompt = """You are a medical/healthcare query analysis assistant. 
Analyze queries and extract structured information about intent, medical domains, complexity, and key terms.

Return a JSON object with:
- intent: Primary intent (diagnosis, treatment, research, consultation, education, regulation, compliance, etc.)
- domains: Array of medical domains (cardiology, oncology, neurology, immunology, etc.)
- complexity: Complexity level (low, medium, high)
- keywords: Key medical terms and concepts (array of strings)
- medicalTerms: Specific medical terminology found in the query (array of strings)
- confidence: Confidence score (0-1) in your analysis

Focus on medical/healthcare context. If not medical, still analyze for general intent.
Be precise with domain classification and keyword extraction."""

            # Call OpenAI API
            response = self.openai_client.chat.completions.create(
                model="gpt-4-turbo-preview",
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": query}
                ],
                response_format={"type": "json_object"},
                temperature=0.3,
                max_tokens=500,
                timeout=30.0  # 30 second timeout
            )
            
            # Parse response
            content = response.choices[0].message.content
            if not content:
                raise ValueError("Empty response from LLM")
            
            analysis_data = json.loads(content)
            
            # Build response with validation
            analysis = QueryAnalysisResponse(
                intent=analysis_data.get("intent", "general"),
                domains=analysis_data.get("domains", []),
                complexity=analysis_data.get("complexity", "medium"),
                keywords=analysis_data.get("keywords", []),
                medical_terms=analysis_data.get("medicalTerms", []),
                confidence=min(max(float(analysis_data.get("confidence", 0.7)), 0.0), 1.0),
                correlation_id=correlation_id
            )
            
            # Log success with metrics
            duration_ms = (datetime.now() - start_time).total_seconds() * 1000
            logger.info(
                "query_analysis_completed",
                operation="analyzeQuery",
                operation_id=operation_id,
                correlation_id=correlation_id,
                intent=analysis.intent,
                domains=analysis.domains,
                complexity=analysis.complexity,
                confidence=analysis.confidence,
                duration_ms=duration_ms,
                token_usage={
                    "prompt_tokens": response.usage.prompt_tokens if hasattr(response, 'usage') else 0,
                    "completion_tokens": response.usage.completion_tokens if hasattr(response, 'usage') else 0,
                    "total_tokens": response.usage.total_tokens if hasattr(response, 'usage') else 0
                }
            )
            
            return analysis
            
        except json.JSONDecodeError as e:
            logger.error(
                "query_analysis_json_parse_failed",
                operation="analyzeQuery",
                operation_id=operation_id,
                correlation_id=correlation_id,
                error=str(e),
                content_preview=content[:200] if 'content' in locals() else None
            )
            # Fallback to basic analysis
            return self._fallback_analysis(query, correlation_id)
            
        except Exception as e:
            duration_ms = (datetime.now() - start_time).total_seconds() * 1000
            logger.error(
                "query_analysis_failed",
                operation="analyzeQuery",
                operation_id=operation_id,
                correlation_id=correlation_id,
                error=str(e),
                error_type=type(e).__name__,
                duration_ms=duration_ms
            )
            # Fallback to basic analysis
            return self._fallback_analysis(query, correlation_id)
    
    def _fallback_analysis(
        self,
        query: str,
        correlation_id: Optional[str] = None
    ) -> QueryAnalysisResponse:
        """
        Fallback analysis when LLM call fails
        
        Provides basic analysis using keyword extraction and pattern matching.
        
        Args:
            query: User query text
            correlation_id: Optional correlation ID
            
        Returns:
            QueryAnalysisResponse with fallback values
        """
        logger.warn(
            "using_fallback_analysis",
            correlation_id=correlation_id,
            query_preview=query[:100]
        )
        
        # Basic keyword extraction
        keywords = [
            word.lower() 
            for word in query.split() 
            if len(word) > 3 and word.isalnum()
        ][:10]
        
        # Simple domain detection from keywords
        medical_keywords = {
            'cardiology': ['heart', 'cardiac', 'cardiovascular', 'ecg', 'echocardiography'],
            'oncology': ['cancer', 'tumor', 'oncology', 'chemotherapy', 'radiation'],
            'neurology': ['brain', 'neurological', 'stroke', 'epilepsy', 'dementia'],
            'immunology': ['immune', 'allergy', 'immunotherapy', 'autoimmune'],
            'endocrinology': ['diabetes', 'thyroid', 'hormone', 'metabolic'],
        }
        
        detected_domains = []
        for domain, domain_keywords in medical_keywords.items():
            if any(keyword in query.lower() for keyword in domain_keywords):
                detected_domains.append(domain)
        
        # Determine intent from keywords
        intent_keywords = {
            'diagnosis': ['diagnose', 'symptom', 'sign', 'condition', 'disease'],
            'treatment': ['treat', 'therapy', 'medication', 'intervention', 'cure'],
            'research': ['study', 'trial', 'research', 'evidence', 'publication'],
            'consultation': ['advice', 'recommendation', 'consult', 'opinion'],
            'education': ['learn', 'understand', 'explain', 'information', 'guide'],
            'regulation': ['fda', 'ema', 'regulatory', 'compliance', 'guideline'],
        }
        
        detected_intent = 'general'
        for intent, intent_keywords_list in intent_keywords.items():
            if any(keyword in query.lower() for keyword in intent_keywords_list):
                detected_intent = intent
                break
        
        return QueryAnalysisResponse(
            intent=detected_intent,
            domains=detected_domains[:3] if detected_domains else [],
            complexity='medium',
            keywords=keywords[:5],
            medical_terms=keywords[:3],
            confidence=0.5,
            correlation_id=correlation_id
        )

    async def select_agents_hybrid(
        self,
        query: str,
        tenant_id: str,
        mode: str = "mode2",
        max_agents: int = 3,
        min_confidence: float = 0.70,
        correlation_id: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Select agents using GraphRAG 3-method hybrid search.

        This is the primary agent selection method that uses:
        - PostgreSQL full-text search (30% weight)
        - Pinecone vector search (50% weight)
        - Neo4j graph traversal (20% weight)

        Combined using weighted Reciprocal Rank Fusion (RRF).

        Args:
            query: User query text
            tenant_id: Tenant identifier for multi-tenant isolation
            mode: Ask Expert mode (mode1-mode4)
            max_agents: Maximum number of agents to return
            min_confidence: Minimum confidence threshold
            correlation_id: Optional correlation ID for tracing

        Returns:
            List of selected agents with confidence scores
        """
        logger.info(
            "hybrid_agent_selection_started",
            query_preview=query[:100],
            tenant_id=tenant_id,
            mode=mode,
            max_agents=max_agents,
            use_graphrag=self.use_graphrag,
            correlation_id=correlation_id
        )

        # Use GraphRAG if enabled and available
        if self.use_graphrag and self.graphrag_selector:
            try:
                agents = await self.graphrag_selector.select_agents(
                    query=query,
                    tenant_id=tenant_id,
                    mode=mode,
                    max_agents=max_agents,
                    min_confidence=min_confidence
                )

                logger.info(
                    "hybrid_agent_selection_completed",
                    method="graphrag",
                    agents_selected=len(agents),
                    correlation_id=correlation_id
                )

                return agents

            except Exception as e:
                logger.error(
                    "graphrag_selection_failed_fallback",
                    error=str(e),
                    error_type=type(e).__name__,
                    correlation_id=correlation_id
                )
                # Fall through to legacy method

        # Fallback to legacy embedding-based selection
        return await self._select_agents_legacy(
            query=query,
            tenant_id=tenant_id,
            max_agents=max_agents,
            correlation_id=correlation_id
        )

    async def _select_agents_legacy(
        self,
        query: str,
        tenant_id: str,
        max_agents: int = 3,
        correlation_id: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Legacy agent selection using query analysis and simple scoring.
        Used as fallback when GraphRAG is unavailable.
        """
        # Analyze the query
        analysis = await self.analyze_query(query, correlation_id)

        if not self.supabase_client:
            logger.warning(
                "legacy_selection_unavailable",
                reason="no_supabase_client",
                correlation_id=correlation_id
            )
            return []

        # Get active agents for tenant - CRITICAL: Use limit to prevent fetching 100+ agents
        # Fetch 3x max_agents to have enough candidates for scoring
        candidate_limit = max_agents * 3
        agents = await self.supabase_client.get_all_agents(
            tenant_id=tenant_id,
            status="active",
            limit=candidate_limit
        )

        # Score agents based on query analysis
        scored_agents = []
        for agent in agents:
            score = 0.5  # Base score
            agent_domains = agent.get("knowledge_domains", [])

            # Domain match scoring
            for domain in analysis.domains:
                if domain.lower() in [d.lower() for d in agent_domains]:
                    score += 0.3

            # Tier weighting
            tier = agent.get("tier", 2)
            if analysis.complexity == "high" and tier == 3:
                score += 0.2
            elif analysis.complexity == "medium" and tier == 2:
                score += 0.1

            scored_agents.append({
                **agent,
                "fused_score": min(score, 1.0),
                "confidence": {
                    "overall": min(score * 100, 95),
                    "methods_found": 1,
                    "breakdown": {"legacy": score * 100}
                }
            })

        # Sort by score and return top-k
        scored_agents.sort(key=lambda x: x["fused_score"], reverse=True)

        logger.info(
            "legacy_agent_selection_completed",
            agents_selected=len(scored_agents[:max_agents]),
            correlation_id=correlation_id
        )

        return scored_agents[:max_agents]

    async def select_multiple_experts_diverse(
        self,
        query: str,
        max_agents: int = 3,
        tenant_id: Optional[str] = None,
        mode: str = "mode4"
    ) -> List[Dict[str, Any]]:
        """
        Select multiple diverse experts for a query (Mode 4 compatibility).

        Uses GraphRAG hybrid selection if available, otherwise falls back
        to legacy selection.

        Args:
            query: User query to analyze
            max_agents: Maximum number of agents to select
            tenant_id: Optional tenant ID for filtering
            mode: Ask Expert mode (default mode4)

        Returns:
            List of selected agent dictionaries
        """
        if not tenant_id:
            tenant_id = self.settings.default_tenant_id if hasattr(self.settings, 'default_tenant_id') else "platform"

        return await self.select_agents_hybrid(
            query=query,
            tenant_id=tenant_id,
            mode=mode,
            max_agents=max_agents,
            min_confidence=0.60  # Lower threshold for Mode 4 diversity
        )


# ============================================================================
# SERVICE FACTORY
# ============================================================================

_agent_selector_service: Optional[AgentSelectorService] = None


def get_agent_selector_service(
    supabase_client: Optional[SupabaseClient] = None
) -> AgentSelectorService:
    """
    Get or create agent selector service instance (singleton pattern)

    Args:
        supabase_client: Optional Supabase client for dependency injection.
                        If provided and the singleton exists without a client,
                        the client will be injected.

    Returns:
        AgentSelectorService instance
    """
    global _agent_selector_service

    if _agent_selector_service is None:
        _agent_selector_service = AgentSelectorService(
            supabase_client=supabase_client
        )
    elif supabase_client is not None and _agent_selector_service.supabase_client is None:
        # Inject supabase_client if the singleton was created without one
        _agent_selector_service.supabase_client = supabase_client
        # Also update the graphrag_selector if it exists
        if _agent_selector_service.graphrag_selector:
            _agent_selector_service.graphrag_selector.supabase = supabase_client

    return _agent_selector_service


def reset_agent_selector_service():
    """Reset the singleton for testing purposes."""
    global _agent_selector_service
    _agent_selector_service = None
